chain/consensus/iface.go

```diff
@@ -18,12 +18,35 @@ import (
 )
 
 type Consensus interface {
-	ValidateBlock(ctx context.Context, b *types.FullBlock) (err error)
+	// ValidateBlockHeader is called by peers when they receive a new block through the network.
+	//
+	// This is a fast sanity-check validation performed by the PubSub protocol before delivering
+	// it to the syncer. It checks that the block has the right format and it performs
+	// other consensus-specific light verifications like ensuring that the block is signed by
+	// a valid miner, or that it includes all the data required for a full verification.
 	ValidateBlockHeader(ctx context.Context, b *types.BlockHeader) (rejectReason string, err error)
+
+	// ValidateBlock is called by the syncer to determine if to accept a block or not.
+	//
+	// It performs all the checks needed by the syncer to accept
+	// the block (signature verifications, VRF checks, message validity, etc.)
+	ValidateBlock(ctx context.Context, b *types.FullBlock) (err error)
+
+	// IsEpochBeyondCurrMax is used to configure the fork rules for longest-chain
+	// consensus protocols.
 	IsEpochBeyondCurrMax(epoch abi.ChainEpoch) bool
 
+	// CreateBlock implements all the logic required to propose and assemble a new Filecoin block.
+	//
+	// This function encapsulate all the consensus-specific actions to propose a new block
+	// such as the ordering of transactions, the inclusion of consensus proofs, the signature
+	// of the block, etc.
```

Comment:
Who calls `CreateBlock` and when? Can the consensus layer count on this being called by Lotus code, the contract being that on each call of `CreateBlock` a new block has to be proposed (and (much later) delivered and returned from `CreateBlock`)? Is the returned block supposed to be ready for execution?

Currently we're working with the idea that the consensus algorithm chooses itself when it proposes new blocks. I.e., it has access to the mempool and pulls requests ("messages") from it whenever it wants. It then "announces" the delivered blocks to Lotus (still Eudico in our case) through an API, and Lotus (Eudico) then takes care of the execution.

I don't have a principal problem if the triggers for proposing the block came from outside (from the point of view of the consensus protocol) instead, I'd just like to know from where and how. I guess it's the miner that calls `CreateBlock`, right? In this case the miner will also need to run consensus-specific logic (as we do in Eudico now), and maybe it will even end up completely circumventing this function...